geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java

```diff
@@ -248,16 +290,23 @@ void saveLines(List<LineGeoData> linesGeoData) {
         // TODO filter by country
         Map<String, LineGeoData> linesGeoDataDb = lineCustomRepository.getLines();
 
-        List<LineGeoData> linesGeoDb = network.getLineStream()
+        List<Line> lines = network.getLineStream()
                 .filter(line -> countries.isEmpty()
                         || line.getTerminal1().getVoltageLevel().getSubstation().getCountry().map(countries::contains).isPresent()
                         || line.getTerminal2().getVoltageLevel().getSubstation().getCountry().map(countries::contains).isPresent())
-                .map(line -> linesGeoDataDb.get(line.getId()))
-                .filter(Objects::nonNull)
                 .collect(Collectors.toList());
-
+        // we also want the destination substation (so we add the neighbouring country)
+        Set<Country> countryAndNextTo =
+            lines.stream().flatMap(line -> line.getTerminals().stream().map(term -> term.getVoltageLevel().getSubstation().getNullableCountry()).filter(Objects::nonNull))
+            .collect(Collectors.toSet());
+        Map<String, SubstationGeoData> substationGeoDataDb = getSubstationMap(network, countryAndNextTo);
+        List<LineGeoData> lineGeoData = lines.stream().map(line -> getLineGeoDataWithEndSubstations(linesGeoDataDb, substationGeoDataDb, line)).collect(Collectors.toList());
         LOGGER.info("{} lines read from DB in {} ms", linesGeoDataDb.size(),  stopWatch.getTime(TimeUnit.MILLISECONDS));
 
-        return linesGeoDb;
+        return lineGeoData;
+    }
+
+    private Map<String, SubstationGeoData> getSubstationMap(Network network, Set<Country> countries) {
```

Comment:
not sure if it's better to load all the substations from all the potential countries, or only the ones we actually use..

Reply:
we can filter by substations and country 
SELECT * FROM geo_data.substations WHERE country in ('SC', 'HT') and id in ('SMURF','GARGA', 'AZRAE')

Reply:
Is 'in' clause a really good idea when we have a large number of element?

Reply:
the clause IN is generally bad with performance, it may be better to use multiple query